### PR TITLE
Added parceler `@Extra` and `@InstanceState` support

### PR DESCRIPTION
**Pull request by johncarl81**
This PR offers [Parceler](http://parceler.org) integration which wraps and unwraps beans annotated with `@Parcel` transparently when they are used in `@Extra` or `@InstanceState` aspects.  This is a rework of this previous PR: #977

### CODE REVIEW

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/parceler/ParcelerBean.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	public String getName() {
+		return name;
+	}
+
+	public int getAge() {
+		return age;
+	}
+}
```
Comment (dodgex) [inline AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/parceler/ParcelerBean.java]: no newline at end of file

Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/parceler/ParcelerBean.java]: Oops.. done.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java [hunk 2]
```diff
@@ -135,21 +138,36 @@ public JInvocation getSuperPutExtraInvocation(TypeMirror elementType, JVar extra
...
+				TypeMirror stringType = elementUtils.getTypeElement(STRING).asType();
+				if (!typeUtils.isSubtype(elementType, stringType)) {
+					extraParameterArg = cast(environment.getClasses().SERIALIZABLE, extraParameterArg);
+				}
 			}
 		}
 		return _super().invoke("extra").arg(extraKeyField).arg(extraParameterArg);
 	}
 
+	public boolean isAnnotatedWith(TypeMirror typeMirror, String qualifedName) {
```
Comment (dodgex) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: you already added this method in `AnnotationHelper` any chance to use that one?

Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: I certainly could, but I wasn't sure how to get ahold of AnnotationHelper other than making `isAnnotatedWith` static.  What do you suggest?

File: AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/parceler/ParcelerBean.java (hunk 1 above)
Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations-test/src/main/java/org/androidannotations/test/parceler/ParcelerBean.java]: #1560

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java (hunk 2 above)
Comment (dodgex) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: I'd pass the `AnnotationHelper` to the intent builder.

Add it to the constructor, then update the `ActivityIntentBuilder` and `ServiceIntentBuilder` to constructors and thier instantiations. where they are instantiated there is `annotationHelper` field from parent class available.

but i think you should do that in a separate commit before doing the work that adds the parceler support

Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: Got it... done.

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: I think you still forgot to remove the method from here.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java [hunk 3]
```diff
@@ -172,6 +173,10 @@ public BundleHelper(AndroidAnnotationsEnvironment environment, TypeMirror elemen
 			if (isTypeParcelable(type)) {
 				methodNameToSave = "put" + "Parcelable";
 				methodNameToRestore = "get" + "Parcelable";
+			} else if (annotationHelper.isAnnotatedWith(type, "org.parceler.Parcel")) {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: `Parcels.wrap/unwrap` method [can handle](http://stackoverflow.com/questions/29674199/use-android-annotations-with-parceler/29680080#comment47641892_29691052) `List<AnnotatedWithParcel>` AFAIK, or even more. We should also allow those.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/FragmentArgHandler.java [hunk 4]
```diff
@@ -106,7 +106,10 @@ private void createBuilderInjectionMethod(Element element, EFragmentHolder holde
 		JClass paramClass = codeModelHelper.typeMirrorToJClass(type);
 
 		JMethod method = builderClass.method(PUBLIC, holder.narrow(builderClass), fieldName);
-		JVar arg = method.param(paramClass, fieldName);
+		JExpression arg = method.param(paramClass, fieldName);
+		if (bundleHelper.isParcelerBean()) {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/FragmentArgHandler.java]: We should add this to `BundleHelper` in some way, to remove code duplication.

**Comment by WonderCsabo** [comment]
Sorry, but i am not sure the refactor with passing the `AnnotationHelper` instance around is needed. You could just create a new `AnnoationHelper` in the intent builder constructor.

```java
annotationHelper = new AnnotationHelper(environment);
```

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/FragmentArgHandler.java (hunk 4 above)
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/FragmentArgHandler.java]: Maybe we could just create the whole expression with `BundleHelper` just like when we restore, instead of only get the method name, and leave the caller (AnnotationHandler) to create the whole saving expression.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java (hunk 2 above)
Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: It got missed in the rebase.. should be removed now.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/FragmentArgHandler.java (hunk 4 above)
Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/handler/FragmentArgHandler.java]: Sounds good... done.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java (hunk 3 above)
Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: Parceler can also handle boxed primitives in collections:
`List<Integer>`, `Map<Integer, String>`
A mix:
`Map<String, AnnotatedWithParcel>`
And nested collections:
`Map<List<String>, List<AnnotatedWithParcel>>`
Would you like these cases handled as well?

**Comment by yDelouis** [comment]
Could you please update the javadoc of the annotations `@Extra`, `@FragmentArg` and `@SavedInstanceState` to specify that these annotations are compatible with classes annotated with `@Parceler` ?

**Comment by yDelouis** [comment]
And I think you forgot to handle inner `Extra` annotations like `@OnActivityResult.Extra`. You have to modify the class `ExtraParameterHandler`.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java (hunk 3 above)
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: If you can add it, it would be great. However we should also check that we have `Parceler` on the classpath (when the user adds a `List` of primitives for examples, we cannot be sure because we are not checking any annotation).

**Comment by WonderCsabo** [comment]
@yDelouis Those extra annotations are already handled by the `BundleHelper`.

**Comment by WonderCsabo** [comment]
@johncarl81 this branch also has to be rebased.

**Comment by yDelouis** [comment]
@WonderCsabo, you're right, sorry.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java (hunk 3 above)
Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: I've been going with the assumption that if you annotate your bean with `@Parcel` that would mean Parceler is on the classpath.  Perhaps we should just support collections with supported types and at least one bean annotated with `@Parcel`... would that be acceptable instead of checking explicitly that Parceler is on the classpath?

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: @johncarl81 it is very easy to check it is on the classpath: `elementUtils.getTypeElement("org.parceler.Parcer") != null`. We use this type of check for lots of things throughout AA.

Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: Ok, that will work nicely, I'll move forward with that then.

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: Great!

Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: I'm finding that matching types following the critieria we discussed may be a bit overzelous.   For instance, this is matching the `ArrayList<String> listExtra` in `ExtraInjectedActivity`.  Should we be serializing this through Parceler?

Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: I guess previously `ArrayList<String>` was being handled as a `Serializable`.  Maybe it is better to use Parcelable to serialize it.

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: No, i think we should only serialize those objects through Parceler which are not in plain Android.

Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: Agreed... I think I have a solution, pushing shortly.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/ValidatorHelper.java [hunk 5]
```diff
@@ -546,8 +546,10 @@ public void canBePutInABundle(Element element, ElementValidation valid) {
 			if (typeMirror.getKind() != TypeKind.NONE) {
 				TypeMirror parcelableType = annotationHelper.typeElementFromQualifiedName(CanonicalNameConstants.PARCELABLE).asType();
 				TypeMirror serializableType = annotationHelper.typeElementFromQualifiedName("java.io.Serializable").asType();
-				if (!annotationHelper.isSubtype(typeMirror, parcelableType) && !annotationHelper.isSubtype(typeMirror, serializableType)) {
-					valid.addError("Unrecognized type. Please let your attribute be primitive or implement Serializable or Parcelable");
+				if (!annotationHelper.isSubtype(typeMirror, parcelableType)
+						&& !annotationHelper.isSubtype(typeMirror, serializableType)
+						&& !annotationHelper.isParcelType(typeMirror)) {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/ValidatorHelper.java]: I just realized we have a problem here. We only allow `Parcelable`, `Serializable` (and now Parceler supported) types. But that will leave out most of [these types for example](https://github.com/johncarl81/androidannotations/blob/parceler/AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java#L47). We have to fix this later...

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/AnnotationHelper.java [hunk 6]
```diff
@@ -55,6 +66,40 @@
 
 	public static final String DEFAULT_FIELD_NAME_VALUE = "value";
 	public static final String DEFAULT_FIELD_NAME_RESNAME = "resName";
+	private static final String PARCEL_ANNOTATION = "org.parceler.Parcel";
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/AnnotationHelper.java]: Can you move all Parceler-related things to another helper class (which can extend `AnnotationHelper`)?

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/AnnotationHelper.java]: I would move `PARCEL_ANNOTATION` to `CanonicalNameConstants` and use that constant everywhere.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/AnnotationHelper.java [hunk 7]
```diff
@@ -95,6 +140,47 @@ public TypeElement typeElementFromQualifiedName(String qualifiedName) {
...
+						&& (!root || genericsSize > 0)) {
+					for (int i = 0; i < genericsSize; i++) {
+						genericsMatch &= isParcelType(declaredType.getTypeArguments().get(i), false);
+					}
+
+					return genericsMatch;
+				}
+			}
+		}
+		return false;
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/AnnotationHelper.java]: Very nice algorithm.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/internal/helper/AndroidManifestFinderTest.java [hunk 8]
```diff
@@ -122,4 +122,4 @@ private void createFile(File file) throws IOException {
 			fail("Could not creates test manifest" + file);
 		}
 	}
-}
\ No newline at end of file
+}
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/test/java/org/androidannotations/internal/helper/AndroidManifestFinderTest.java]: This is done on develop, please revert.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java [hunk 9]
```diff
@@ -135,16 +139,18 @@ public JInvocation getSuperPutExtraInvocation(TypeMirror elementType, JVar extra
...
-				if (!typeUtils.isSubtype(elementType, stringType)) {
-					extraParameterArg = cast(environment.getClasses().SERIALIZABLE, extraParameterArg);
-				}
-			} else {
+			if (typeUtils.isSubtype(elementType, parcelableType)) {
 				TypeMirror serializableType = elementUtils.getTypeElement(SERIALIZABLE).asType();
 				if (typeUtils.isSubtype(elementType, serializableType)) {
 					extraParameterArg = cast(environment.getClasses().PARCELABLE, extraParameterArg);
 				}
+			} else if (!BundleHelper.METHOD_SUFFIX_BY_TYPE_NAME.containsKey(elementType.toString()) && annotationHelper.isParcelType(elementType)) {
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: I just realized that `METHOD_SUFFIX_BY_TYPE_NAME` is not containing everything which is supported by `Bundle`.

File: AndroidAnnotations/androidannotations-core/androidannotations-test/pom.xml [hunk 10]
```diff
@@ -82,6 +82,17 @@
 			<version>140</version>
 			<scope>test</scope>
 		</dependency>
+		<dependency>
+			<groupId>org.parceler</groupId>
+			<artifactId>parceler</artifactId>
+			<version>1.0.3</version>
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations-test/pom.xml]: Can you extract the version to a property?

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java [hunk 11]
```diff
@@ -135,16 +139,18 @@ public JInvocation getSuperPutExtraInvocation(TypeMirror elementType, JVar extra
...
+				extraParameterArg = environment.getCodeModel().ref("org.parceler.Parcels").staticInvoke("wrap").arg(extraParameterArg);
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: Please update the comment `// Cast to Parcelable or Serializable if needed`. This is what i hate about comments, they rot...

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/ValidatorHelper.java (hunk 5 above)
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/ValidatorHelper.java]: I am stupid, ignore this.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java [hunk 12]
```diff
@@ -225,6 +231,10 @@ public JExpression getExpressionToRestoreFromBundle(JClass variableClass, JExpre
 			expressionToRestore = JExpr.invoke(bundle, methodNameToRestore).arg(extraKey);
 		}
 
+		if (parcelerBean) {
+			expressionToRestore = environment.getCodeModel().ref("org.parceler.Parcels").staticInvoke("unwrap").arg(expressionToRestore);
```
Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/helper/BundleHelper.java]: Can you extract `"org.parceler.Parcels"` to `CanonicalNameConstants`?

**Comment by WonderCsabo** [comment]
This is OK, however i realized `METHOD_SUFFIX_BY_TYPE_NAME` is outdated and we also do not support `SparceArray<T extends Parcelable>`.

File: AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java (hunk 9 above)
Comment (johncarl81) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: This code is begging for a more unified approach.  For example, check out the way I did this 
in Parceler:
[Generators](https://github.com/johncarl81/parceler/blob/master/parceler/src/main/java/org/parceler/internal/Generators.java)
[Generators setup](https://github.com/johncarl81/parceler/blob/master/parceler/src/main/java/org/parceler/internal/ParcelerModule.java#L219-L258)
[Generators validation](https://github.com/johncarl81/parceler/blob/master/parceler/src/main/java/org/parceler/internal/ParcelableAnalysis.java#L575)
[Generators usage - generates read code](https://github.com/johncarl81/parceler/blob/master/parceler/src/main/java/org/parceler/internal/ParcelableGenerator.java#L284)
and in Transfuse:
[Bundle Property Repository](https://github.com/johncarl81/transfuse/blob/master/transfuse/src/main/java/org/androidtransfuse/analysis/repository/BundlePropertyBuilderRepository.java)
[Bundle Property Setup](https://github.com/johncarl81/transfuse/blob/master/transfuse/src/main/java/org/androidtransfuse/config/TransfuseAndroidModule.java#L219-L234)
[Bundle Property validation](https://github.com/johncarl81/transfuse/blob/master/transfuse/src/main/java/org/androidtransfuse/gen/variableBuilder/ExtraInjectionNodeBuilder.java#L93)
[Bundle Property usage - generates write code](https://github.com/johncarl81/transfuse/blob/master/transfuse/src/main/java/org/androidtransfuse/gen/IntentFactoryStrategyGenerator.java#L153)

Comment (WonderCsabo) [inline AndroidAnnotations/androidannotations-core/androidannotations/src/main/java/org/androidannotations/internal/core/helper/IntentBuilder.java]: Totally agree, the current code is just too complicated...

I also thought about creating a provider thing, which we discussed in your
original PR.
That way we could completely split off Parceler support to a plugin,
instead of adding Parceler-related
code directly to AA.

Your code is very clean and robust, congratulations!

**Comment by WonderCsabo** [comment]
Can you rebase this? We switched our codemodel dependency, which uses some new class names and namespace.

**Comment by johncarl81** [comment]
Sure!  Done.

**Comment by WonderCsabo** [comment]
Thanks!

@yDelouis can we merge this?

**Comment by yDelouis** [comment]
Javadoc for `@Extra`, `@FragmentArg` and `@SavedInstanceState` are still missing. But except, that, it is okay.

**Comment by WonderCsabo** [comment]
And from even more annotations, you are right.

**Comment by johncarl81** [comment]
Ok, let me know if you like that language.

By the way, if you guys would like great looking javadocs with less ceremony I'd recommend [Asciidoclet](http://asciidoctor.org/news/2013/06/03/asciidoclet-announcement/)

**Comment by WonderCsabo** [comment]
We also have to add info for:

* `OnActivityResult.Extra`
* `Receiver.Extra`
* `ReceiverAction.Extra`
* `ServiceAction.Extra` (actually we already miss parameter type info on that :( )

**Comment by johncarl81** [comment]
I don't see `ServiceAction.Extra'...

**Comment by WonderCsabo** [comment]
Sorry. Just ServiceAction.

**Comment by johncarl81** [comment]
Alright, I think I got them all... let me know if there's anything else.

**Comment by johncarl81** [comment]
Ping

**Comment by WonderCsabo** [comment]
Sorry for not adding feedback, i am very busy at work :(

I think this can be merged. @yDelouis ?

**Comment by yDelouis** [comment]
I think so ! Thanks !

**Comment by johncarl81** [comment]
Fantastic!

Thanks @WonderCsabo for your help on this effort, both on this PR and your attention to Eclipse support in Parceler.  Without your diligence this feature would have never reached this point.

**Comment by WonderCsabo** [comment]
@johncarl81 i hope i could help, and thank you for implementing this in AA, it was a quick task.

@yDelouis we have to update the wiki. I think we can add this information in the third party integration section. Is that OK?

**Comment by WonderCsabo** [comment]
@johncarl81 i created a [wiki page](https://github.com/excilys/androidannotations/wiki/ParcelerIntegration) for Parceler integration, i hope everything is correct.

**Comment by johncarl81** [comment]
Looks spot on @WonderCsabo.

**Comment by WonderCsabo** [comment]
Great, thanks for checking.

**Comment by WonderCsabo** [comment]
@johncarl81 FYI i tried this in Gradle, and the `JCodeModel#ref` method throwed an `NoClassDefFoundError` about `android.os.Parcelable` when you are referencing the `Parcels` utility class. So i changed to use `JCodeModel#directClass` instead 12342808.

**Comment by johncarl81** [comment]
Ah, good catch.
